pkg/cloudprovider/providers/gce/gce_loadbalancer_external_test.go

```diff
@@ -574,3 +578,412 @@ func TestLoadBalancerWrongTierResourceDeletion(t *testing.T) {
+	assert.Empty(t, ipAddress)
+	assert.NotNil(t, err)
+
+	// When portRange != the forwardingRule port range.
+	wrongPorts := []v1.ServicePort{apiService.Spec.Ports[0]}
+	wrongPorts[0].Port = wrongPorts[0].Port + 1
+	exists, needsUpdate, ipAddress, err = gce.forwardingRuleNeedsUpdate(lbName, region, apiService.Spec.LoadBalancerIP, wrongPorts)
+	assert.True(t, exists)
+	assert.True(t, needsUpdate)
+	assert.Equal(t, ipAddr, ipAddress)
+	assert.NoError(t, err)
+
+	// When the ports protocol does not equal the ForwardingRuel IP Protocol.
+	wrongPorts = []v1.ServicePort{apiService.Spec.Ports[0]}
+	wrongPorts[0].Protocol = v1.ProtocolUDP
+	exists, needsUpdate, ipAddress, err = gce.forwardingRuleNeedsUpdate(lbName, region, apiService.Spec.LoadBalancerIP, wrongPorts)
+	assert.True(t, exists)
+	assert.True(t, needsUpdate)
+	assert.Equal(t, ipAddr, ipAddress)
+	assert.NoError(t, err)
+
+	// Basic workflow.
+	exists, needsUpdate, ipAddress, err = gce.forwardingRuleNeedsUpdate(lbName, region, apiService.Spec.LoadBalancerIP, apiService.Spec.Ports)
+	assert.True(t, exists)
+	assert.False(t, needsUpdate)
+	assert.Equal(t, ipAddr, ipAddress)
+	assert.NoError(t, err)
+}
+
+func TestTargetPoolNeedsRecreation(t *testing.T) {
+	gce, err := fakeGCECloud()
+	require.NoError(t, err)
+
+	lbName := cloudprovider.GetLoadBalancerName(apiService)
+	nodes, err := createAndInsertNodes(gce, []string{nodeName})
+	require.NoError(t, err)
+	hostNames := nodeNames(nodes)
+	hosts, err := gce.getInstancesByNames(hostNames)
+
+	var instances []string
+	for _, host := range hosts {
+		instances = append(instances, host.makeComparableHostPath())
+	}
+	pool := &compute.TargetPool{
+		Name:            lbName,
+		Description:     fmt.Sprintf(`{"kubernetes.io/service-name":"%s"}`, serviceName),
+		Instances:       instances,
+		SessionAffinity: translateAffinityType(v1.ServiceAffinityNone),
+	}
+	err = gce.CreateTargetPool(pool, region)
+	require.NoError(t, err)
+
+	c := gce.c.(*cloud.MockGCE)
+	c.MockTargetPools.GetHook = mock.GetTargetPoolInternalErrHook
+	exists, needsRecreation, err := gce.targetPoolNeedsRecreation(lbName, region, v1.ServiceAffinityNone)
+	assert.True(t, exists)
+	assert.False(t, needsRecreation)
+	require.NotNil(t, err)
+	assert.True(t, strings.HasPrefix(err.Error(), "error getting load balancer's target pool:"))
+	c.MockTargetPools.GetHook = nil
+
+	exists, needsRecreation, err = gce.targetPoolNeedsRecreation(lbName, region, v1.ServiceAffinityClientIP)
+	assert.True(t, exists)
+	assert.True(t, needsRecreation)
+	assert.NoError(t, err)
+
+	exists, needsRecreation, err = gce.targetPoolNeedsRecreation(lbName, region, v1.ServiceAffinityNone)
+	assert.True(t, exists)
+	assert.False(t, needsRecreation)
+	assert.NoError(t, err)
+}
+
+func TestFirewallNeedsUpdate(t *testing.T) {
+	gce, err := fakeGCECloud()
+	require.NoError(t, err)
+	status, err := createExternalLoadBalancer(gce)
+	require.NotNil(t, status)
+	require.NoError(t, err)
+	svcName := "/"
```

Comment:
minor: give it an actual name (e.g. "bar-svc")?

Reply:
Changed. The service definition defers from its actual name.